### ISSUE TITLE
Elasticsearch 400 error 

### ISSUE THREAD

**Issue by dannyshox** [heroku/predictionio-buildpack#39]
Hi,

Getting this 400 error on training ( bin/pio train -- --driver-memory 8G )

`Exception in thread "main" org.elasticsearch.client.ResponseException: PUT http://127.0.0.1:9200/pio_meta/_mapping/apps: HTTP/1.1 400 Bad Request
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"mapper [id] cannot be changed from type [text] to [keyword]"}],"type":"illegal_argument_exception","reason":"mapper [id
] cannot be changed from type [text] to [keyword]"},"status":400}`

The ES logs show : 

`[2017-06-19T10:25:24,241][WARN ][o.e.c.r.a.DiskThresholdMonitor] [RpmdHBT] high disk watermark [90%] exceeded on [RpmdHBTMQfOowYrOf0fgDQ][RpmdHBT][/Users/doshaughnessy/Desktop/Apps/Predictio
nIO/apache-predictionio-0.11.0-incubating/PredictionIO-0.11.0-incubating/apps/predictionio-engine-ur/PredictionIO-dist/vendors/elasticsearch/data/nodes/0] free: 34.5gb[7.4%], shards will be
relocated away from this node`

`[2017-06-19T10:25:24,242][INFO ][o.e.c.r.a.DiskThresholdMonitor] [RpmdHBT] rerouting shards: [high disk watermark exceeded on one or more nodes]`

`[2017-06-19T10:25:25,735][DEBUG][o.e.a.a.i.m.p.TransportPutMappingAction] [RpmdHBT] failed to put mappings on indices [[[pio_meta/SnfZ-1HnS0KzMSObghhEbw]]], type [apps]
java.lang.IllegalArgumentException: mapper [id] cannot be changed from type [text] to [keyword]`

This only happens in my local environment . It builds, trains & deploys OK on Heroku.

**Comment by mars** [comment]
Elasticsearch throws that error sometimes, enough that I recognize it! It's a rough edge that I want to figure out how to avoid. 

What does `bin/pio status` return?

I've had luck fixing it by resetting storage; easiest way to ensure clean databases is:

Open `./PredictionIO/conf/pio-env.sh` and change the storage prefixes to something like `pio_meta_local_ur`, `pio_event_local_ur`, & `pio_model_local_ur`. This will give you a clean slate, and it ensures the databases don't conflict with other engines.

Then, set it up again per the [instructions](https://github.com/heroku/predictionio-engine-ur/blob/master/README.md#user-content-import-sample-data).

**Comment by mars** [comment]
Filed a bug w/ Apache PredictionIO
https://issues.apache.org/jira/plugins/servlet/mobile#issue/PIO-96

**Comment by dannyshox** [comment]
Thanks Mars.

pio status :

`[INFO] [Management$] Inspecting PredictionIO...
[INFO] [Management$] PredictionIO 0.11.0-SNAPSHOT is installed at /Users/doshaughnessy/Desktop/Apps/PredictionIO/apache-predictionio-0.11.0-incubating/PredictionIO-0.11.0-incubating/apps/pre
dictionio-engine-ur/PredictionIO-dist
[INFO] [Management$] Inspecting Apache Spark...
[INFO] [Management$] Apache Spark is installed at /Users/doshaughnessy/Desktop/Apps/PredictionIO/apache-predictionio-0.11.0-incubating/PredictionIO-0.11.0-incubating/apps/predictionio-engine
-ur/PredictionIO-dist/vendors/spark-hadoop
[INFO] [Management$] Apache Spark 2.1.0 detected (meets minimum requirement of 1.3.0)
[INFO] [Management$] Inspecting storage backend connections...
[INFO] [Storage$] Verifying Meta Data Backend (Source: ELASTICSEARCH)...
[INFO] [Storage$] Verifying Model Data Backend (Source: PGSQL)...
[INFO] [Storage$] Verifying Event Data Backend (Source: PGSQL)...
[INFO] [Storage$] Test writing to Event Store (App Id 0)...
[INFO] [Management$] Your system is all ready to go.`

I ran into another issue after the above. The data wasn't being loaded into the app.

I think the issue was at the start of the import-events bash script :

![screen shot 2017-06-20 at 08 52 35](https://user-images.githubusercontent.com/4091125/27322457-1b7e5122-5596-11e7-8c45-b22eb2a8569f.png)

I have a global 'pio' in my bash_profile so it was using this and loading the data into an old app I had called ur !

Working fine now once I commented that out.

**Comment by dannyshox** [comment]
... and adding the line `PIO_BIN=bin/pio` instead

**Comment by mars** [comment]
You definitely hit a rough edge of PIO local development. The strategy of having a single `pio` installation on the PATH seems innocent enough, but it leads to complexity and corruption of data when different engines each need different storage config. These issues went 💥 when we started working with the Universal Recommender, because our engines before that used the same all-Postgres storage config.

I just opened a [pull request adding guardrails to `bin/local/setup`](https://github.com/heroku/predictionio-buildpack/pull/40), preventing accidental use of an existing `pio` command. Does this seem reasonable? I'm trying to imagine if this would turn into a pain in the neck for PredictionIO developers that also have a use-case for global `pio`.
